common/lib/client/auth.js

```diff
@@ -534,6 +519,144 @@ var Auth = (function() {
 		}
 	};
 
+	Auth.prototype._saveBasicOptions = function(defaultTokenParams, authOptions) {
+		this.method = 'basic';
+		this.key = authOptions.key;
+		this.basicKey = toBase64(authOptions.key);
+		this.tokenParams = defaultTokenParams || {};
+		this.authOptions = authOptions || {};
+		delete this.authOptions.force;
+		if('clientId' in authOptions) {
+			this._userSetClientId(authOptions.clientId);
+		}
+	}
+
+	Auth.prototype._saveTokenOptions = function(tokenParams, authOptions) {
+		this.method = 'token';
+
+		this.tokenParams = tokenParams || this.tokenParams || {};
+
+		/* If an authOptions object is passed in that contains new auth info (ie
+		* isn't just {force: true} or something), it becomes the new default, with
+		* the exception of the force attribute (RSA10g), which is set anew on each
+		* call to authorise (defaulting to false) */
+		this.force = false;
+		if(authOptions) {
+			this.force = authOptions.force;
+
+			if(containsAuthInfo(authOptions)) {
+				this.authOptions = authOptions || {};
```

Comment:
> Of course this is why it was a bad idea all along to have authorise() persist the overrides, and a setOptions(), which fully supersedes the previously set options, would have been more intuitive IMO :)

I am not sure how it changes anything really.  Either `authorise` persists new changes or does not persist, but whether it merges is not relevant as far as I can tell because we'll still need to decide on merging options or not.

> Merging (and then persisting) is inconsistent with the other APIs and doesn't let you unset things once they're set in java and similar languages. Unsetting them by including authURL: undefined etc is a bit horrible anyway.

Yes, it is in typed langs.

> So, if we're not going to have a setOptions() which explicitly and fully supersedes the set options, then what @SimonWoolf has done here (which is piecemeal and arbitrary) may well deal with the typical use-cases as well as possible.

How?  If you provide a set of options to `authorise` regardless of `setOptions` existing will either merge with existing options or not.  

I assume a merge will happen regardless based on the default values anyway i.e. whatever you provide will merge with `TokenParam` and `AuthOption` defaults.

>>  believe we have merged in all other implementations

> It's not as if it's a huge change, so unfortunately I think that's not a good enough reason by itself to go further along the road of making an inferior API.

It's big enough that we'll need to change tests in every library now, so I agree going down the wrong road is not necessarily the right approach, but this is not trivial given the number of client libraries and the test coverage that exists in each.

The problem now though if we did go down this route is that if you wanted to reauth without changing any options and thus set the `force: true` key/value of `AuthOptions`, how will the client library now know that you the rest of the `AuthOptions` should be ignored.  There is no non-ugly way to do that, especially in typed languages.  Also, if a user currently provides `{}` as an AuthOption arg in JS, it will reset `AuthOption` to empty, which IMHO, could be a little confusing for users.